src/ShellCheck/Parser.hs

```diff
@@ -136,6 +136,31 @@ almostSpace =
         char c
         return ' '
 
+withNextId :: Monad m => SCParser m (Id -> b) -> SCParser m b
+withNextId p = do
+    start <- getPosition
+    id <- createId
+    fn <- p
```

Comment:
What about the parsers that skip trailing whitespace?

The second `T_NormalWord` for `echo foo      bar` would here be marked with a span of 10 instead of of 3

Reply:
I haven't tried or looked very deeply, but we can probably avoid this by calling withNextId inside functions that read whitespace, rather than wrapping them. e.g. in readCmdWord, we can do withNextId readNormalWord; spacing, instead of withNextId readCmdWord. Do you think this will work? (I will get to trying it this weekend :)

Reply:
I think that could lead to a lot of issues in the near term. Given that most of the parsers only use the `id` when returning a value, maybe it should be constructed in two phases and without the higher order function, e.g.:

```
parse = do
    start <- getNextId
    whatever
    span <- endId start
    spacing
    return $ T_Stuff span [...] 
```

This has the flexibility of the manual approach and the safety of typechecking

Reply:
Good point. I implemented this in a series of commits (trying to make changes small).

e496b41 essentially reverts withNextId
0c459ae adds a function that given a id (representing start pos), updates the position map to set the end pos to the current position (maybe need a better name)
3848788 is the demo using this new function when reading dollar variable

I looked at the output using `shellcheck -f json test.sh`, with
```
$ cat test.sh
#!/bin/bash
echo      $1       $2
```

Reply:
Oh and the output of `shellcheck -f json test.sh` is:
```
[
   {
      "line" : 2,
      "file" : "test2.sh",
      "endLine" : 2,
      "column" : 11,
      "endColumn" : 13,
      "level" : "info",
      "message" : "Double quote to prevent globbing and word splitting.",
      "code" : 2086
   },
   {
      "code" : 2086,
      "message" : "Double quote to prevent globbing and word splitting.",
      "level" : "info",
      "column" : 20,
      "endLine" : 2,
      "endColumn" : 22,
      "file" : "test2.sh",
      "line" : 2
   }
]
```
looks fine, endColumn is exclusive

Reply:
It still looks unchecked to me. Doesn't it still compile if you forget to add `endPosOfStartId id`?

Reply:
Sorry I probably misunderstood it. Referring to your example previously:
```
parse = do
    start <- getNextId
    whatever
    span <- endPosOfStartId start
    spacing
    return $ T_Stuff span [...] 
```
So change `endPosOfStartId` to return an Id? Or do you mean for `span` to be a different type?

Reply:
It would be convenient if `endPosOfStartId` returned an `Id` for compatibility. The main point is that if you accidentally forget to mark the start position or the end position, the code should not compile. 

This could be one example implementation with different names:

```
data IncompleteInterval = IncompleteInterval Position

startSpan :: m IncompleteInterval
startSpan = IncompleteInterval <$> getPosition

endSpan :: IncompleteInterval -> m Id
endSpan (IncompleteInterval start) = do
  end <- getPosition
  id <- createId
  registerStartAndEnd (start, end) id
  return id

readThing = do
  start <- startSpan
  whatever
  id <- endSpan start
  return $ T_Stuff id
```

In this case (at least if `createId` was hidden), you would not be able to return an AST mode without making sure both the start position and the end position is set. If you forget to do either, the compiler will remind you. This ensures that no refactoring or new code causes any nodes to have missing end positions, since this can be hard to notice during testing.

Some kind of `createZeroWidthId` could be useful while refactoring for code that doesn't mark end positions yet.

Reply:
Understood:
- followed your suggested API to create a startSpan and endSpan
- updated the existing implementations to use this new API
- `getNextId` now creates an empty span
- make the endPos non-optional in the map